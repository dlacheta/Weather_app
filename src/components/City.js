import React from 'react';
import PropTypes from 'prop-types';

const City = ({
  city: {
    name,
    country,
    population,
    coord: { lat, lon },
  },
}) => (
  <div className="card">
    <div className="card-body">
      <h4 className="card-title">{`${name}, ${country}`}</h4>

      <div className="d-flex justify-content-between">
        <p>{`Population: ${population}`}</p>
        <p>{`Coordinates: Latitute: ${lat} Longitude: ${lon}`}</p>
      </div>
    </div>
  </div>
);

City.propTypes = {
  city: PropTypes.shape({
    name: PropTypes.string,
    country: PropTypes.string,
    population: PropTypes.number,
    coord: PropTypes.shape({
      lat: PropTypes.number,
      lon: PropTypes.number,
    }),
  }).isRequired,
};

export default City;
