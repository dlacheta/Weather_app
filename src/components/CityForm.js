import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getWeather, toogleLoading } from '../actions/rootActions';

export const CityForm = ({ toogleLoadingDispatch, getWeatherDispatch }) => {
  const submitForm = (event) => {
    event.preventDefault();
    if (event.target.city_name.value) {
      toogleLoadingDispatch();
      getWeatherDispatch(event.target.city_name.value);
    }
  };
  return (
    <form onSubmit={submitForm}>
      <div className="input-group mb-3">
        <input name="city_name" type="text" className="form-control" placeholder="City" />
        <div className="input-group-append">
          <button className="btn btn-outline-dark" type="submit" id="button-addon2">
            Submit
          </button>
        </div>
      </div>
    </form>
  );
};

CityForm.propTypes = {
  toogleLoadingDispatch: PropTypes.func.isRequired,
  getWeatherDispatch: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  getWeatherDispatch: city => dispatch(getWeather(city)),
  toogleLoadingDispatch: () => dispatch(toogleLoading()),
});

export default connect(
  null,
  mapDispatchToProps,
)(CityForm);
