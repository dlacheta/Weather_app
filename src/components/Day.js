import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';

const Day = ({ today, weather }) => (
  <div className="card">
    <div className="card-body">
      <h5 className="card-title">
        {today
          ? `Today (${moment(weather[0].dt_txt).format('DD - MMM - YYYY')})`
          : `${moment(weather[0].dt_txt).format('dddd')} (${moment(weather[0].dt_txt).format(
            'DD - MMM - YYYY',
          )})`}
      </h5>
      <ul className="list-group">
        {weather.map(time => (
          <li className="d-flex list-group-item align-middle" key={time.dt}>
            <div className="time">{moment(time.dt_txt).format('HH:mm')}</div>
            <div className="icon">
              <img
                src={`http://openweathermap.org/img/w/${time.weather[0].icon}.png`}
                alt={time.weather[0].description}
              />
            </div>
            <div className="temp">
              {time.main.temp}
              <sup>o</sup>
C
            </div>
            <div className="hpa">{`${time.main.pressure} hPa`}</div>
            <div className="wind">{time.wind.speed}</div>
            <div className="wind-arrow">
              <i className="material-icons" style={{ transform: `rotate(${time.wind.deg}deg)` }}>
                arrow_right_alt
              </i>
            </div>
          </li>
        ))}
      </ul>
    </div>
  </div>
);

Day.propTypes = {
  today: PropTypes.bool.isRequired,
  weather: PropTypes.arrayOf(
    PropTypes.shape({
      dt: PropTypes.number,
      dt_txt: PropTypes.string,
      weather: PropTypes.arrayOf(
        PropTypes.shape({
          icon: PropTypes.string,
          description: PropTypes.string,
        }),
      ),
      main: PropTypes.shape({
        temp: PropTypes.number,
        pressure: PropTypes.number,
      }),
      wind: PropTypes.shape({
        speed: PropTypes.number,
        deg: PropTypes.number,
      }),
    }),
  ).isRequired,
};

export default Day;
