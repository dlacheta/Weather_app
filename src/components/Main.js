import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Day from './Day';
import City from './City';
import CityForm from './CityForm';
import capitalizeFirstLetter from '../func/capitalizeFirstLetter';

export const Main = ({
  city, weather, loadingWeather, error,
}) => (
  <div>
    <div className="jumbotron jumbotron-fluid">
      <div className="container">
        <h1 className="display-4">Weather app</h1>
        <div className="lead">
          Please enter a name of the city to get 5 day forecast:
          <CityForm />
        </div>
        {city.name ? <City city={city} /> : capitalizeFirstLetter(error)}
      </div>
    </div>
    <div className="d-flex flex-row bd-highlight mb-3 flex-wrap">
      {city.name
        ? Object.keys(weather).map((day, index) => (
          <Day key={day} today={index === 0} weather={weather[day]} />
        ))
        : null}
    </div>
    {loadingWeather ? (
      <div className="loading-container">
        <div className="loading" />
      </div>
    ) : null}
  </div>
);

Main.propTypes = {
  city: PropTypes.shape({
    name: PropTypes.string,
    country: PropTypes.string,
    population: PropTypes.number,
    coord: PropTypes.shape({
      lat: PropTypes.number,
      lon: PropTypes.number,
    }),
  }),
  weather: PropTypes.arrayOf(
    PropTypes.arrayOf(
      PropTypes.shape({
        dt: PropTypes.number,
        dt_txt: PropTypes.string,
        weather: PropTypes.arrayOf(
          PropTypes.shape({
            icon: PropTypes.string,
            description: PropTypes.string,
          }),
        ),
        main: PropTypes.shape({
          temp: PropTypes.number,
          pressure: PropTypes.number,
        }),
        wind: PropTypes.shape({
          speed: PropTypes.number,
          deg: PropTypes.number,
        }),
      }),
    ),
  ),
  loadingWeather: PropTypes.bool,
  error: PropTypes.string,
};

Main.defaultProps = {
  city: {
    name: '',
    country: '',
    population: 0,
    coord: {
      lat: 0,
      lon: 0,
    },
  },
  weather: [
    [
      {
        dt: 0,
        dt_txt: '',
        weather: [
          {
            icon: '',
            description: '',
          },
        ],
        main: {
          temp: 0,
          pressure: 0,
        },
        wind: {
          speed: 0,
          deg: 0,
        },
      },
    ],
  ],
  loadingWeather: false,
  error: '',
};

const mapStateToProps = state => ({
  city: state.city,
  weather: state.weather,
  loadingWeather: state.loadingWeather,
  error: state.error,
});

export default connect(mapStateToProps)(Main);
