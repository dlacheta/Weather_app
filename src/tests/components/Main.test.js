import React from 'react';
import { shallow } from 'enzyme';
import { Main } from '../../components/Main';

it('Main component renders correctly', () => {
  const wrapper = shallow(<Main />);
  expect(wrapper).toMatchSnapshot();
});
