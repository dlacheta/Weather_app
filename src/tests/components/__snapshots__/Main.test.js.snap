// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Main
    city={
      Object {
        "coord": Object {
          "lat": 0,
          "lon": 0,
        },
        "country": "",
        "name": "",
        "population": 0,
      }
    }
    error=""
    loadingWeather={false}
    weather={
      Array [
        Array [
          Object {
            "dt": 0,
            "dt_txt": "",
            "main": Object {
              "pressure": 0,
              "temp": 0,
            },
            "weather": Array [
              Object {
                "description": "",
                "icon": "",
              },
            ],
            "wind": Object {
              "deg": 0,
              "speed": 0,
            },
          },
        ],
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="jumbotron jumbotron-fluid"
        >
          <div
            className="container"
          >
            <h1
              className="display-4"
            >
              Weather app
            </h1>
            <div
              className="lead"
            >
              Please enter a name of the city to get 5 day forecast:
              <Connect(CityForm) />
            </div>
            
          </div>
        </div>,
        <div
          className="d-flex flex-row bd-highlight mb-3 flex-wrap"
        />,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="container"
          >
            <h1
              className="display-4"
            >
              Weather app
            </h1>
            <div
              className="lead"
            >
              Please enter a name of the city to get 5 day forecast:
              <Connect(CityForm) />
            </div>
            
          </div>,
          "className": "jumbotron jumbotron-fluid",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="display-4"
              >
                Weather app
              </h1>,
              <div
                className="lead"
              >
                Please enter a name of the city to get 5 day forecast:
                <Connect(CityForm) />
              </div>,
              "",
            ],
            "className": "container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Weather app",
                "className": "display-4",
              },
              "ref": null,
              "rendered": "Weather app",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Please enter a name of the city to get 5 day forecast:",
                  <Connect(CityForm) />,
                ],
                "className": "lead",
              },
              "ref": null,
              "rendered": Array [
                "Please enter a name of the city to get 5 day forecast:",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            "",
          ],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
          "className": "d-flex flex-row bd-highlight mb-3 flex-wrap",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="jumbotron jumbotron-fluid"
          >
            <div
              className="container"
            >
              <h1
                className="display-4"
              >
                Weather app
              </h1>
              <div
                className="lead"
              >
                Please enter a name of the city to get 5 day forecast:
                <Connect(CityForm) />
              </div>
              
            </div>
          </div>,
          <div
            className="d-flex flex-row bd-highlight mb-3 flex-wrap"
          />,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="container"
            >
              <h1
                className="display-4"
              >
                Weather app
              </h1>
              <div
                className="lead"
              >
                Please enter a name of the city to get 5 day forecast:
                <Connect(CityForm) />
              </div>
              
            </div>,
            "className": "jumbotron jumbotron-fluid",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h1
                  className="display-4"
                >
                  Weather app
                </h1>,
                <div
                  className="lead"
                >
                  Please enter a name of the city to get 5 day forecast:
                  <Connect(CityForm) />
                </div>,
                "",
              ],
              "className": "container",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Weather app",
                  "className": "display-4",
                },
                "ref": null,
                "rendered": "Weather app",
                "type": "h1",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Please enter a name of the city to get 5 day forecast:",
                    <Connect(CityForm) />,
                  ],
                  "className": "lead",
                },
                "ref": null,
                "rendered": Array [
                  "Please enter a name of the city to get 5 day forecast:",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {},
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": "div",
              },
              "",
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
            "className": "d-flex flex-row bd-highlight mb-3 flex-wrap",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
